src/diff.rs

```diff
@@ -182,91 +201,683 @@ fn diff_attributes(change_list: &mut ChangeListBuilder, old: &[Attribute], new:
+    // old children and we're finished.
+    if shared_prefix_count == new.len() {
+        // Same as above.
+        debug_assert!(pushed);
+        change_list.pop_push_next_sibling();
+        change_list.remove_self_and_next_siblings();
+        return KeyedPrefixResult::Finished;
+    }
+
+    if pushed {
+        debug_assert!(shared_prefix_count > 0);
+        change_list.pop();
+    }
+    KeyedPrefixResult::MoreWorkToDo(shared_prefix_count)
+}
+
+// The most-general, expensive code path for keyed children diffing.
+//
+// We find the longest subsequence within `old` of children that are relatively
+// ordered the same way in `new` (via finding a longest-increasing-subsequence
+// of the old child's index within `new`). The children that are elements of
+// this subsequence will remain in place, minimizing the number of DOM moves we
+// will have to do.
+//
+// Upon entry to this function, the stack must be:
+//
+//     [... parent]
+//
+// Upon exit from this function, it will be restored to that same state.
+fn diff_keyed_middle(
+    cached_set: &CachedSet,
+    change_list: &mut ChangeListBuilder,
+    registry: &mut EventsRegistry,
+    old: &[Node],
+    new: &[Node],
+    cached_roots: &mut FxHashSet<CacheId>,
+    shared_prefix_count: usize,
+    shared_suffix_count: usize,
+    old_shared_suffix_start: usize,
+) {
+    // Should have already diffed the shared-key prefixes and suffixes.
+    debug_assert_ne!(new.first().map(|n| n.key()), old.first().map(|o| o.key()));
+    debug_assert_ne!(new.last().map(|n| n.key()), old.last().map(|o| o.key()));
+
+    // The algorithm below relies upon using `u32::MAX` as a sentinel
+    // value, so if we have that many new nodes, it won't work. This
+    // check is a bit academic (hence only enabled in debug), since
+    // wasm32 doesn't have enough address space to hold that many nodes
+    // in memory.
+    debug_assert!(new.len() < u32::MAX as usize);
+
+    // Map from each `old` node's key to its index within `old`.
+    let mut old_key_to_old_index = FxHashMap::default();
+    old_key_to_old_index.reserve(old.len());
+    old_key_to_old_index.extend(old.iter().enumerate().map(|(i, o)| (o.key(), i)));
+
+    // The set of shared keys between `new` and `old`.
+    let mut shared_keys = FxHashSet::default();
+    // Map from each index in `new` to the index of the node in `old` that
+    // has the same key.
+    let mut new_index_to_old_index = Vec::with_capacity(new.len());
+    new_index_to_old_index.extend(new.iter().map(|n| {
+        let key = n.key();
+        if let Some(&i) = old_key_to_old_index.get(&key) {
+            shared_keys.insert(key);
+            i
+        } else {
+            u32::MAX as usize
+        }
+    }));
+
+    // If none of the old keys are reused by the new children, then we
+    // remove all the remaining old children and create the new children
+    // afresh.
+    if shared_suffix_count == 0 && shared_keys.is_empty() {
```

Comment:
This condition is throwing me a bit for a spin, I am naively expecting it to *just* be checking the `shared_keys` set, but I can't quite figure out why `shared_suffix_count` is factored in here?

Should `shared_suffix_count == 0` only be used to optimize which opcodes are emitted? I'd expect that we'd still have some optimized path for shared `shared_suffix_count` is nonzero but there's no shared keys

Reply:
Your intuition that there should be a fast path regardless of shared_suffix_count's value is correct. The reason it doesn't exist is simply that we don't currently have instructions/functions for mass deleting and inserting in the middle of the children, only all of the children, or the latter half of the children.

It is somethign that I feel can be a follow up PR.